### PR TITLE
refactor: make the main HTTP router not dependent on the active GraphQL server

### PR DESCRIPTION
**Pull request by clayne11**
## Motivation and Context

By decoupling the top level router's HTTP router and the currently active GraphQL server, we enable the router to be decoupled from running the HTTP server.

This separates concerns between configuring and executing the HTTP handler from running the server, and enables consumers to run their own HTTP handler while the internal logic can hot-swap when a new graph configuration is available.

Some consumers may already have an HTTP server running and simply want the router to be attached on a given path. This is the first PR along the path of enabling this capability.

This PR is purely an internal refactor. There are zero logical or API changes.

### CODE REVIEW

File: router/core/router.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ type (
 		configPoller configpoller.ConfigPoller
 		selfRegister selfregister.SelfRegister
 
-		registrationInfo *nodev1.RegistrationInfo
+		publicKey *ecdsa.PublicKey
```
Comment (clayne11) [inline router/core/router.go:122]: `registrationInfo` was only ever used to create the `publicKey`. We eagerly create the key instead on the first `router.Start`.

File: router/core/router.go [hunk 2]
```diff
@@ -135,7 +135,11 @@ type (
 		rootContext       context.Context
 		rootContextCancel func()
 		routerConfig      *nodev1.RouterConfig
-		healthChecks      health.Checker
```
Comment (clayne11) [inline router/core/router.go:138]: There's no reason for the healthChecks to be on the currently active server vs the main router. I pulled it up to the main Router instance.

File: router/core/router.go [hunk 3]
```diff
@@ -871,6 +856,42 @@ func (r *Router) newServer(ctx context.Context, routerConfig *nodev1.RouterConfi
 	return ro, nil
 }
 
+func (r *Router) buildHTTPRouter() http.Handler {
```
Comment (clayne11) [inline router/core/router.go:859]: This HTTP router now has 0 dependencies on the active server or its config. In the next PR I will modify the API to only call this once and move the HTTP server into `cmd.Main`.

File: router/core/router.go [hunk 4]
```diff
@@ -326,9 +336,67 @@ func NewRouter(opts ...Option) (*Router, error) {
 		r.logger.Warn("Development mode enabled. This should only be used for testing purposes")
 	}
 
+	r.httpRouter = r.buildHTTPRouter()
+
 	return r, nil
 }
 
+func (r *Router) buildHTTPRouter() http.Handler {
```
[the remark this answers is not in the document]
Comment (StarpTech) [reply router/core/router.go:320]: What is your motivation? Do you want to run the router along another app or just have flexibility? Are you aware that the router shutdown itself on schema updates?

**Review by StarpTech — APPROVED** [review]
I like the changes. I re-tested the relevant places. LGTM

**Comment by StarpTech** [comment]
Unfortunately, we found a [breaking change](https://github.com/wundergraph/cosmo/pull/347/files#diff-aa58824b56f499a6a7d39dd7d905130a1aed0683be3aa6826a63dfd5d5047edcR362) in this PR. The playground was exposed on all methods which caused conflicts when using `/` as graphql path. During working on the fix, I recognized that the code is harder to reason about compared to the previous code. We need to find another way to decouple it. We have plans to make the router working with AWS lambda. I think this is the right moment to overhaul it.

**Pull request merged**